### DISCUSSION TITLE
How to Change Table of Contents Placement in Quarto Book

### DISCUSSION THREAD

**Discussion by TomBener** [quarto-dev/quarto-cli#9839] [Q&A] [answered]
I’m working on a Quarto thesis book project and need to place the Table of Contents (TOC) _after_ the preface content, which is written in `index.qmd` (preface, acknowledgements, abstract etc.). I found a solution for PDF-only using LaTeX customization (from [this Stack Overflow question](https://stackoverflow.com/questions/73304608/how-can-i-change-the-placement-of-table-of-contents-in-a-quarto-book-project)), but this approach requires to move contents in `index.qmd` to LaTeX, which is not compatible with other formats, such as HTML and DOCX.

What is the best approach to achieve this consistent TOC placement for multiple formats?

Thank you!

**Comment by mcanouil** [comment]
Use `toc-location` as described in the documentation.
The reference pages are intended to be as exhaustive as possible regarding the options for the main formats.
It's a good place to start looking before going a bit randomly on the Internet.


If the option does not fit your needs, then it's going to be very tricky to achieve what you want and more advanced (if even possible).

**Reply by TomBener** [reply]
Thanks, but `toc-location` is not what I need. I want to change the order of TOC in LaTeX.

**Reply by mcanouil** [reply]
Then you need to tweak the template or partials.
See <https://quarto.org/docs/journals/templates.html>.

Alternatively, you can use `format: latex` and post process the LaTeX code.

**Reply by TomBener** [reply]
I've tried template and template partials. But the order of the toc and body is fixed in the template:

```
$if(toc)$
$if(toc-title)$
\renewcommand*\contentsname{$toc-title$}
$endif$
$if(beamer)$
\begin{frame}[allowframebreaks]
$if(toc-title)$
  \frametitle{$toc-title$}
$endif$
  \tableofcontents[hideallsubsections]
\end{frame}
$else$
{
$if(colorlinks)$
\hypersetup{linkcolor=$if(toccolor)$$toccolor$$else$$endif$}
$endif$
\setcounter{tocdepth}{$toc-depth$}
\tableofcontents
}
$endif$
$endif$
$if(lof)$
\listoffigures
$endif$
$if(lot)$
\listoftables
$endif$
$if(linestretch)$
\setstretch{$linestretch$}
$endif$
$if(has-frontmatter)$
\mainmatter
$endif$
$body$
```

It seems the best way is by post-processing the LaTeX code.

**Reply by mcanouil** [reply]
Modify the template or the partial and change the order, and use that modified version.
That's the whole point of template/partials, i.e., being able to replace everything or only parts of the default.

**Reply by TomBener** [reply]
`$body$` Represents everything converted from Markdown, which includes preface, acknowledgements, abstract and subsequent chapters. How can I insert toc between them by changing the template?

**Reply by mcanouil** [reply]
Use YAML or LUA filter to have a "preface" content that you can place where you want in your template.
There is also the "before-body.tex()" partial.

For example, there is a LUA that allow to write the abstract as a section https://github.com/pandoc-ext/abstract-section

**Reply by TomBener** [reply]
@mcanouil Thanks!

**Comment by TomBener** [comment, marked as answer]
I came up with an easy method to adjust the order of TOC in LaTeX:

1. set `toc: false` in `_quarto.yml` for the PDF format.
2. In `index.qmd`, write the raw LaTeX code for printing TOC after the preface. For example:

````md
# Praface {-}

# Acknowledgements {-}

# Abstract {-}

```{=latex}
\clearpage
{
\hypersetup{linkcolor=blue}
\setcounter{tocdepth}{3}
\tableofcontents
}
```
````

The generated LaTeX code would be:

```tex

\bookmarksetup{startatroot}

\chapter*{Praface}\label{praface}
\addcontentsline{toc}{chapter}{Praface}

\markboth{Praface}{Praface}

\bookmarksetup{startatroot}

\chapter*{Acknowledgements}\label{acknowledgements}
\addcontentsline{toc}{chapter}{Acknowledgements}

\markboth{Acknowledgements}{Acknowledgements}

\bookmarksetup{startatroot}

\chapter*{Abstract}\label{abstract}
\addcontentsline{toc}{chapter}{Abstract}

\markboth{Abstract}{Abstract}

\clearpage
{
\hypersetup{linkcolor=blue}
\setcounter{tocdepth}{3}
\tableofcontents
}
```
